draft-ietf-gnap-core-protocol.md

```diff
@@ -2368,108 +2370,199 @@ unknown_request
-note that the RO MAY open the URL on a separate device from the client instance
-itself. The user code URL MUST be accessible from an HTTP GET request,
-and MUST be protected by HTTPS or equivalent means.
+When the RO enters this code at the user code URI,
+the AS MUST uniquely identify the pending request that the code was associated
+with. If the AS does not recognize the entered code, the interaction component MUST
+display an error to the user. If the AS detects too many unrecognized codes
+enter attempts, the interaction component SHOULD display an error to the user.
 
-While it is common for the RO to be the same party as the end-user, since
-the client instance has to communicate the user code to someone, there are
-cases where the end-user and RO are separate parties and the authorization
-happens asynchronously.
+The client instance MUST NOT modify the URI when launching it,
+in particular the client instance MUST NOT add any parameters to the URI.
+The user code URI MUST be reachable from the end user's browser, though
+the URI is usually be opened on a separate device from the client instance
+itself. The URI MUST be accessible from an HTTP GET
+request and MUST be protected by HTTPS or equivalent means.
 
+### Interaction through an Application URI {#interaction-app}
 
-## Interaction through an Application URI {#interaction-app}
+When the client instance is directed to launch an application through the
+["app"](#response-interact-app) mode, the client launches the
+URL as appropriate to the system. The means by which the AS and the
+launched application communicate with each other and perform any
+of the required actions are out of scope for this specification.
 
-When the client instance successfully launches an application through the
-["app" mode](#response-interact-app), the AS
-interacts with the RO through that application to authenticate the
-user as the RO and gather their consent. The details of this
-interaction are out of scope for this specification.
+## Post-Interaction Completion {#interaction-finish}
 
-## Post-Interaction Completion {#interaction-finalize}
-
-Upon completing an interaction with the RO, if a ["callback"](#response-interact-callback) mode is
-available with the current request, the AS MUST follow the appropriate
-method at the end of interaction to allow the client instance to continue. If
-this mode is not available, the AS SHOULD instruct the RO to
-return to their client instance upon completion. Note that these steps
-still take place in most error cases, such as when the RO has denied
-access. This pattern allows the client instance to potentially recover from the error
-state without restarting the request from scratch by modifying its
-request or providing additional information directly to the AS.
+If an interaction ["finish"](#response-interact-finish) method is
+associated with the current request, the AS MUST follow the appropriate
```

Comment:
caveat: we said earlier that if an error happens before, the AS MUST not follow

Reply:
This is tricky since for most error conditions we actually do want the AS to follow the "finish" method even in error cases. The only cases where the finish shouldn't be followed are ones where the request or the URI are fishy.